File: src/co/edu/uptc/aj/os/gui/QueuePanel.java
package co.edu.uptc.aj.os.gui;

import java.awt.Color;
import java.util.ArrayList;
import java.util.Enumeration;

import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JTable;
import javax.swing.table.TableColumn;

import co.edu.uptc.aj.os.gui.base.ProcessRenderer;
import co.edu.uptc.aj.os.gui.base.ProcessTableModel;
import co.edu.uptc.aj.os.logic.Process;

@SuppressWarnings("SuspiciousNameCombination")
public class QueuePanel extends JPanel {

    private ProcessTableModel tableModel;
    private JTable table;
    private JScrollPane pane;

    private ArrayList<Process> processes = new ArrayList<>();

    public QueuePanel() {
        setLayout(null);
        setSize(ProcessRenderer.WIDTH, ProcessRenderer.WIDTH);
        setOpaque(false);
        setBackground(Color.decode("#00000000"));

        tableModel = new ProcessTableModel(new Object[]{}, 0);
        table = new JTable(tableModel);

        table.getTableHeader().setDefaultRenderer(new ProcessRenderer());
        table.setOpaque(false);
        table.setBackground(Color.decode("#00000000"));

        pane = new JScrollPane(table);
        pane.setOpaque(false);
        pane.setBackground(Color.decode("#00000000"));
        updatePane();
        add(pane);
    }

    public void setQueueColor(Color color) {
        ProcessRenderer renderer = new ProcessRenderer();
        renderer.setBorderColor(color);
        table.getTableHeader().setDefaultRenderer(renderer);
    }

    public void addProcess(int index, Process process) {
        if (index < 0) return;
        processes.add(index, process);
        updateProcesses();
    }

    private void updateProcesses() {
        clearTable();
        for (Process pro : processes) {
            tableModel.addColumn(pro.getName());
            updatePane();
        }
    }

    public void removeProcess(Process process) {
        Enumeration<TableColumn> cols = table.getColumnModel().getColumns();
        int index = -1;
        int count = 0;
        while (cols.hasMoreElements()) {
            String colTitle = cols.nextElement().getHeaderValue().toString();
            if (process.getName().equalsIgnoreCase(colTitle)) {
                index = count;
                break;
            }
            count += 1;
        }
        if (index < 0) return;
        try {
            processes.remove(index);
        } catch (Exception e) {
        }
        tableModel.removeColumn(index);
        updatePane();
    }

    public void clearTable() {
        for (int i = 0; i < table.getRowCount(); i++) {
            tableModel.removeRow(i);
            i -= 1;
        }
        tableModel.setColumnCount(0);
        tableModel.removeColumn(0);
        updatePane();
    }

    private void updatePane() {
        pane.setBounds(0, 0, (ProcessRenderer.WIDTH * tableModel.getColumnCount()) + 1,
                       ProcessRenderer.WIDTH - 5);
        revalidate();
        repaint();
    }
}

File: src/co/edu/uptc/aj/os/gui/QueuePanel.kt
package co.edu.uptc.aj.os.gui

import co.edu.uptc.aj.os.gui.base.ProcessRenderer
import co.edu.uptc.aj.os.gui.base.ProcessTableModel
import co.edu.uptc.aj.os.gui.base.QueueTable
import co.edu.uptc.aj.os.ignored
import co.edu.uptc.aj.os.logic.Process
import java.awt.Color
import java.awt.Dimension
import java.util.ArrayList
import javax.swing.JPanel
import javax.swing.JScrollPane
import javax.swing.JTable
import kotlin.math.min

class QueuePanel : JPanel() {
    
    private val tableModel: ProcessTableModel
    private val table: QueueTable
    private val pane: JScrollPane
    private val processes = ArrayList<Process>()
    
    init {
        layout = null
        setSize(ProcessRenderer.WIDTH, ProcessRenderer.WIDTH)
        isOpaque = false
        background = Color.decode("#00000000")
        
        tableModel = ProcessTableModel(arrayOf(), 0)
        table = QueueTable(tableModel)
        
        table.tableHeader.defaultRenderer = ProcessRenderer()
        
        table.isOpaque = false
        table.background = Color.decode("#00000000")
        table.preferredScrollableViewportSize =
            Dimension(ProcessRenderer.WIDTH * 10, ProcessRenderer.WIDTH)
        table.autoResizeMode = JTable.AUTO_RESIZE_OFF
        table.rowSelectionAllowed = false
        table.columnSelectionAllowed = false
        table.tableHeader.reorderingAllowed = false
        table.tableHeader.resizingAllowed = false
        
        pane = JScrollPane(
            table, JScrollPane.VERTICAL_SCROLLBAR_AS_NEEDED,
            JScrollPane.HORIZONTAL_SCROLLBAR_ALWAYS)
        pane.isOpaque = false
        pane.background = Color.decode("#00000000")
        updatePane()
        add(pane)
    }
    
    fun setQueueColor(color: Color) {
        val renderer = ProcessRenderer()
        renderer.borderColor = color
        table.tableHeader.defaultRenderer = renderer
    }
    
    fun addProcess(index: Int, process: Process) {
        if (index < 0) return
        processes.add(index, process)
        updateProcesses()
    }
    
    private fun updateProcesses() {
        clearTable()
        for ((name) in processes) {
            ignored { tableModel.addColumn(name) }
            updatePane()
        }
    }
    
    fun removeProcess(process: Process) {
        val cols = table.columnModel.columns
        var index = -1
        var count = 0
        while (cols.hasMoreElements()) {
            val colTitle = cols.nextElement().headerValue.toString()
            if (process.name.equals(colTitle, ignoreCase = true)) {
                index = count
                break
            }
            count += 1
        }
        if (index < 0) return
        ignored { processes.removeAt(index) }
        ignored { tableModel.removeColumn(index) }
        updatePane()
    }
    
    fun clearTable() {
        for (i in 0 until table.rowCount) {
            ignored { tableModel.removeRow(0) }
        }
        tableModel.columnCount = 0
        ignored { tableModel.removeColumn(0) }
        updatePane(true)
    }
    
    private fun updatePane(cleared: Boolean = false) {
        val maxWidth = min(
            ProcessRenderer.WIDTH * tableModel.columnCount + (if (cleared) 0 else 1),
            width)
        pane.setBounds(0, 0, maxWidth, ProcessRenderer.WIDTH - 5)
        table.preferredScrollableViewportSize =
            Dimension(ProcessRenderer.WIDTH * 10, ProcessRenderer.WIDTH)
        table.autoResizeMode = JTable.AUTO_RESIZE_OFF
        ignored { revalidate() }
        ignored { repaint() }
    }
}